R/features.R

```diff
@@ -176,53 +176,95 @@ helper_feature <- function(m, feature_sample) {
 #'
 #' @param data data.table or data.frame. Includes all the features (both factors and possibly others).
 #'
+#' @param newdata data.table or data.frame. New data (features) that has the same
+#' features as \code{data}.
+#'
 #' @return A list that contains the following entries:
 #' \describe{
+#' \item{obj}{List, Contains \describe{
 #' \item{features}{Vector. Contains the names of all the features in \code{data}.}
 #' \item{factor_features}{Vector. Contains the names of all the factors in \code{data}.}
 #' \item{factor_list}{List. Contains each factor and its vector of levels.}
 #' \item{contrasts_list}{List. Contains all the contrasts of the factors.}
+#' }}
+#' \item{model.matrix}{A data.frame containing all of the factors in \code{new_data} as
+#' one-hot encoded variables.}
 #' }
 #'
+#'
 #' @export
 #'
 #' @author Annabelle Redelmeier
 #'
 #' @examples
 #'
 #' data("Boston", package = "MASS")
-#'
-#' x_var <- c("lstat", "chas", "rad", "indus")
+#' x_var <- c("lstat", "rm", "dis", "indus")
 #' y_var <- "medv"
+#' x_train <- as.data.frame(Boston[401:411, x_var])
+#' y_train <- Boston[401:408, y_var]
+#' x_test <- as.data.frame(Boston[1:4, x_var])
 #'
-#' # convert to factors
-#' Boston$rad = as.factor(Boston$rad)
-#' Boston$chas = as.factor(Boston$chas)
-#'
-#' x_train <- Boston[-1:-6, x_var]
-#' y_train <- Boston[-1:-6, y_var]
-#' x_test <- Boston[1:6, x_var]
+#' # convert to factors for illustational purpose
+#' x_train$rm <- factor(round(x_train$rm))
+#' x_test$rm <- factor(round(x_test$rm), levels = levels(x_train$rm))
 #'
-#' dummylist <- make_dummies(data = rbind(x_train, x_test))
+#' dummylist <- make_dummies(data = rbind(x_train, x_test), newdata = x_test)
 #'
-make_dummies <- function(data) {
+make_dummies <- function(data, newdata) {
 
-  contrasts <- features <- factor_features <-  NULL # due to NSE notes in R CMD check
+  contrasts <- features <- factor_features <- model.frame <- model.matrix <- NULL # due to NSE notes in R CMD check
+  #
   if (is.null(colnames(data))) {
     stop("data must have column names.")
   }
-  data <- data.table::as.data.table(as.data.frame(data, stringsAsFactors = FALSE))
+  if (is.null(colnames(newdata))) {
+    stop("newdata must have column names.")
+  }
+  if (is.null(newdata)) {
+    stop("newdata needs to be included.")
+  }
 
+  data <- data.table::as.data.table(as.data.frame(data, stringsAsFactors = FALSE))
```

Comment:
For this and similar places: 

We say that data and newdata must be data.table or data.frame, but still we call as.data.frame, so it could have been matrix. Do we need to handle matrix? If not, we could skip the cumbersome as.data.frame before as.data.table here. We could/should then check that there are no other column classes than is.numeric and is.factor. 

If one passes a data.frame with characters, currently this converts the data.frame characters to a factor with the levels present in that data.frame. If the same or not found in the test data, then the error of not identical levels might be confusing as the user have not specified any levels, our function has created it for them. If we require factor from the beginning we avoid this.

Reply:
This is dealt with.